File: java/src/main/java/com/penapereira/patterns/interpreter/AddExpression.java
package com.penapereira.patterns.interpreter;

/** Nonterminal: expression '+' expression. */
public final class AddExpression implements AbstractExpression {

    private final AbstractExpression left;
    private final AbstractExpression right;

    public AddExpression(AbstractExpression left, AbstractExpression right) {
        this.left = left;
        this.right = right;
    }

    @Override
    public int interpret(Context context) {
        return left.interpret(context) + right.interpret(context);
    }

    @Override
    public String describe() {
        return "(" + left.describe() + " + " + right.describe() + ")";
    }
}

File: typescript/src/interpreter/add-expression.ts
import type { AbstractExpression } from "./abstract-expression.ts";
import type { Context } from "./context.ts";

/** Nonterminal: expression '+' expression. */
export class AddExpression implements AbstractExpression {
  constructor(
    private readonly left: AbstractExpression,
    private readonly right: AbstractExpression,
  ) {}

  interpret(context: Context): number {
    return this.left.interpret(context) + this.right.interpret(context);
  }

  describe(): string {
    return `(${this.left.describe()} + ${this.right.describe()})`;
  }
}
